sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala

```diff
@@ -114,4 +114,197 @@ object EstimationUtils {
     }
   }
 
+  /**
+   * Returns the number of the first bin into which a column values falls for a specified
+   * numeric equi-height histogram.
+   *
+   * @param value a literal value of a column
+   * @param histogram a numeric equi-height histogram
+   * @return the number of the first bin into which a column values falls.
+   */
+
+  def findFirstBinForValue(value: Double, histogram: Histogram): Int = {
+    var binId = 0
+    histogram.bins.foreach { bin =>
+      if (value > bin.hi) binId += 1
+    }
+    binId
+  }
+
+  /**
+   * Returns the number of the last bin into which a column values falls for a specified
+   * numeric equi-height histogram.
+   *
+   * @param value a literal value of a column
+   * @param histogram a numeric equi-height histogram
+   * @return the number of the last bin into which a column values falls.
+   */
+
+  def findLastBinForValue(value: Double, histogram: Histogram): Int = {
+    var binId = 0
+    for (i <- 0 until histogram.bins.length) {
+      if (value > histogram.bins(i).hi) {
+        // increment binId to point to next bin
+        binId += 1
+      }
+      if ((value == histogram.bins(i).hi) && (i < histogram.bins.length - 1)) {
+        if (value == histogram.bins(i + 1).lo) {
```

Comment:
merge two `if`s: if ((value == histogram.bins(i).hi) && (value == histogram.bins(i + 1).lo) && (i < histogram.bins.length - 1))

Reply:
I used two statements instead of one statement is because, when i points to the last bin, this condition "value == histogram.bins(i + 1).lo" may be out of bound.  By separating the conditions into two statements, we can be sure that the out-of-bound error will not happen.

Reply:
By "out of bound", do you mean it exceeds 100 length limit? You can just switch new line after `&&`

Reply:
No.  I meant the upper bound for the array of bins in a histogram.  The default length of the histogram bin array is 254.  When i is equal to 253 (the last bin), then i+1 is 254 leading to out-of-bound error.

Reply:
just move this condition after the length check: 
```
if ((value == histogram.bins(i).hi) && (i < histogram.bins.length - 1) && (value == histogram.bins(i + 1).lo))
```